### PR TITLE
jv-registration-validation HW solution

### PR DESCRIPTION
**Pull request by luk14n**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/InvalidRegistrationDataException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+public class InvalidRegistrationDataException extends RuntimeException {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/InvalidRegistrationDataException.java:3]:
```suggestion
public class RegistrationException extends RuntimeException {
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -5,10 +5,66 @@
...
+    /*
+
+    there is no user with such login in the Storage yet
+    user's login is at least 6 characters
+    user's password is at least 6 characters
+    user's age is at least 18 years old
+
+    */
+    private static final int DEFAULT_USER_LOGIN_SIZE = 6;
+    private static final int DEFAULT_USER_PASSWORD_SIZE = 6;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:17]:
```suggestion
    private static final int MIN_USER_LOGIN_SIZE = 6;
    private static final int MIN_USER_PASSWORD_SIZE = 6;
```

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -5,10 +5,66 @@
...
+    private static final int MINIMUM_USER_AGE = 18;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]:
```suggestion
    private static final int MIN_USER_AGE = 18;
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        registrationService = new RegistrationServiceImpl();
+        user = new User();
+        storageDao = new StorageDaoImpl();
+    }
+
+    @BeforeEach
+    void setUp() {
+        String newUserLogin = "Login123";
+        String newUserPassword = "qwerty123";
+        int newUserAge = 23;
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:30]: remove this local vars

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        user.setAge(newUserAge);
+        user.setLogin(newUserLogin);
+        user.setPassword(newUserPassword);
+    }
+
+    @Test
+    void register_null_user_NotOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:37]:
```suggestion
    void register_nullUser_NotOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        assertThrows(RuntimeException.class,
+                () -> registrationService.register(null));
+    }
+
+    @Test
+    void register_null_login_NotOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:43]:
```suggestion
    void register_nullLogin_NotOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        user.setAge(null);
+        assertThrows(RuntimeException.class,
+                () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_under_age_limit_NotOk() {
+        user.setAge(11);
+        assertThrows(InvalidRegistrationDataException.class,
+                () -> registrationService.register(user));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:91]: image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    @Test
+    void register_empty_login_NotOK() {
+        user.setLogin("");
+        assertThrows(InvalidRegistrationDataException.class,
+                () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_short_password_NotOk() {
+        user.setPassword("123");
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:131]: image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        assertThrows(InvalidRegistrationDataException.class,
+                () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_user_Ok() {
+        User register = new User();
+        register.setLogin("Login123");
+        register.setPassword("Password");
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:140]: image

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    @BeforeEach
+    void setUp() {
+        user.setAge(23);
+        user.setLogin("Login123");
+        user.setPassword("qwerty123");
+    }
+
+    @Test
+    void register_nullUser_NotOk() {
+        assertThrows(RuntimeException.class,
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:35]:
```suggestion
        assertThrows(RegistrationException.class,
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                () -> registrationService.register(null));
+    }
+
+    @Test
+    void register_invalid_login_NotOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:40]:
```suggestion
    void register_invalidLogin_NotOk() {
```
image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                () -> registrationService.register(user));
+        user.setAge(-1);
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+        user.setAge(10);
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+        user.setAge(16);
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:71]: image

**Review by Rommelua — CHANGES_REQUESTED** [review]
Before you create PR with your homework, you need to go through the checklist under the task and correct all the points described there. The mentor will not check the work until the checklist points are corrected

**Comment by luk14n** [comment]
> Before you create PR with your homework, you need to go through the checklist under the task and correct all the points described there. The mentor will not check the work until the checklist points are corrected

Ok, I will pay attention in the future. Forgive my inattention

**Review by Rommelua — CHANGES_REQUESTED** [review]
image

Java Cl checks failed. Before making a Pull Request (or a new commit to an already existing PR), you need to write the mvn clean package command in the terminal and make sure that all checks pass. After that, the mentor will be able to review the homework

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        assertDoesNotThrow(() -> {
+            user.setAge(18);
```
[the remark this answers is not in the document]
Comment (luk14n) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:74]: image
Can you than elaborate, please, more precisely what changes here is required? It seems to me that all edge cases is handled properly

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 14]
```diff
@@ -5,10 +5,69 @@
...
+            throw new
+                    RegistrationException("Login cannot be empty!");
+        }
+        if (user.getAge() == 0) {
+            throw new
+                    RegistrationException("Age must be greater that 0");
+        }
+        if (user.getAge() < 0) {
+            throw new
+                    RegistrationException("Age cannot be negative number!");
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:47]: image
You already have validation

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 15]
```diff
@@ -5,10 +5,69 @@
...
+        if (user.getLogin().isEmpty()) {
+            throw new
+                    RegistrationException("Login cannot be empty!");
+        }
+        if (user.getAge() == 0) {
+            throw new
+                    RegistrationException("Age must be greater that 0");
+        }
+        if (user.getAge() < 0) {
+            throw new
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:46]: You can divide logic of validation age, password, login to helpful private methods

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+    }
+
+    @Test
+    void register_validAge_Ok() {
+        User userAge18 = new User();
+        userAge18.setLogin("UserAge18");
+        userAge18.setPassword("password18");
+        userAge18.setAge(18);
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:83]: Better save magic numbers in constant and correct naming of userValidAge or like this

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        user.setLogin("");
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+        user.setLogin("log");
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+        user.setLogin("login");
+        assertThrows(RegistrationException.class,
+                () -> registrationService.register(user));
+    }
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:54]: Save strings to constant

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by Sanderfuck — APPROVED** [review]

**Pull request closed without merging**